### PR TITLE
[tests] added forms test

### PR DESCRIPTION
**Pull request by radekdoulik**
- to measure UI app startup times

 - added RunUITests task (similar to unInstrumentationTests) to start
   app/activity

  - renamed .*UnitTestApk.* to .*TestApk.* as it handles also the new
   UI test

  - updated ProcessLogcatTiming to be able to process logcat output of
    started Android activity process

### CODE REVIEW

**Comment by radekdoulik** [comment]
the build will fail as it will need a fix from https://github.com/xamarin/xamarin-android/pull/815 to be able to use custom timing definitions

**Comment by jonpryor** [comment]
PR #515 has been merged.

Indentation is...weirdly all over the map. It should be corrected.

File: src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/RunUITests.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+﻿using Microsoft.Build.Framework;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	public class RunUITests : Adb
+	{
+		public                  string              AdbTarget                   { get; set; }
+		public                  string              AdbOptions                  { get; set; }
+
+        [Required]
```
Comment (jonpryor) [inline src/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/RunUITests.cs]: This should be left-aligned with the `public` on the following line.

File: tests/forms/App.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+using Xamarin.Forms;
+
+namespace forms
+{
+    public partial class App : Application
```
Comment (jonpryor) [inline tests/forms/App.xaml.cs]: I take it that the default Xamarin.Forms template uses spaces? Not sure if this needs to be fixed, just seems weird.

File: tests/forms/Droid/MainActivity.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+using Android.Widget;
+using Android.OS;
+
+namespace forms.Droid
+{
+    [Activity(Label = "forms.Droid", Icon = "@drawable/icon", Theme = "@style/MyTheme", MainLauncher = true, ConfigurationChanges = ConfigChanges.ScreenSize | ConfigChanges.Orientation)]
+    public class MainActivity : global::Xamarin.Forms.Platform.Android.FormsAppCompatActivity
+    {
+        bool firstOnCreate = true;
+		bool firstOnStart = true;
```
Comment (jonpryor) [inline tests/forms/Droid/MainActivity.cs]: The `bool` on this and the following line should be indented the same amount as the previous line.

File: tests/forms/Droid/MainActivity.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            base.OnCreate(bundle);
+
+            global::Xamarin.Forms.Forms.Init(this, bundle);
+
+            LoadApplication(new App());
+            if (firstOnCreate) {
+                Console.WriteLine("startup-timing: OnCreate end reached");
+                firstOnCreate = false;
+            }
+		}
```
Comment (jonpryor) [inline tests/forms/Droid/MainActivity.cs]: This line is indented too much; it should be aligned with the `{` for `OnCreate()`.

File: tests/forms/Droid/MainActivity.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        protected override void OnStart()
+        {
+            if (firstOnStart)
+                Console.WriteLine("startup-timing: OnStart reached");
+			base.OnStart();
+            if (firstOnStart) {
+                Console.WriteLine("startup-timing: OnStart end reached");
+                firstOnStart = false;
+            }
+		}
```
Comment (jonpryor) [inline tests/forms/Droid/MainActivity.cs]: This line is indented too much; it should be aligned with the `{` for `OnStart()`.

File: tests/forms/Droid/MainActivity.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
+        {
+            base.OnActivityResult(requestCode, resultCode, data);
+        }
+
+        protected override void OnStart()
+        {
+            if (firstOnStart)
+                Console.WriteLine("startup-timing: OnStart reached");
+			base.OnStart();
```
Comment (jonpryor) [inline tests/forms/Droid/MainActivity.cs]: This line is indented too much; it should be aligned with the `if`, two lines above.

File: tests/forms/Droid/Resources/values/styles.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+		<item name="windowNoTitle">true</item>
+		<!--We will be using the toolbar so no need to show ActionBar-->
+		<item name="windowActionBar">false</item>
+		<!-- Set theme colors from http://www.google.com/design/spec/style/color.html#color-color-palette-->
+		<!-- colorPrimary is used for the default action bar background -->
+		<item name="colorPrimary">#2196F3</item>
+		<!-- colorPrimaryDark is used for the status bar -->
+		<item name="colorPrimaryDark">#1976D2</item>
+		<!-- colorAccent is used as the default value for colorControlActivated
+         which is used to tint widgets -->
```
Comment (jonpryor) [inline tests/forms/Droid/Resources/values/styles.xml]: What made *this* line wrap decision? It's ugly. :-/

File: tests/forms/Services/CloudDataStore.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                items = await Task.Run(() => JsonConvert.DeserializeObject<IEnumerable<Item>>(json));
+            }
+
+            return items;
+        }
+
+        public async Task<Item> GetItemAsync(string id)
+        {
+            if (id != null && CrossConnectivity.Current.IsConnected)
+            {
```
Comment (jonpryor) [inline tests/forms/Services/CloudDataStore.cs]: Ditto.

File: tests/forms/ViewModels/AboutViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace forms
+{
+    public class AboutViewModel : BaseViewModel
+    {
+        public AboutViewModel()
+        {
+            Title = "About";
+
+            OpenWebCommand = new Command(() => Device.OpenUri(new Uri("https://xamarin.com/platform")));
```
Comment (jonpryor) [inline tests/forms/ViewModels/AboutViewModel.cs]: Space ` ` before `(`.

File: tests/forms/ViewModels/BaseViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace forms
+{
+    public class BaseViewModel : INotifyPropertyChanged
+    {
+        public IDataStore<Item> DataStore => DependencyService.Get<IDataStore<Item>>() ?? new MockDataStore();
+
+        bool isBusy = false;
+        public bool IsBusy
+        {
```
Comment (jonpryor) [inline tests/forms/ViewModels/BaseViewModel.cs]: This should be on the previous line.

File: tests/forms/ViewModels/BaseViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            get { return isBusy; }
+            set { SetProperty(ref isBusy, value); }
+        }
+
+        string title = string.Empty;
+        public string Title
+        {
```
Comment (jonpryor) [inline tests/forms/ViewModels/BaseViewModel.cs]: Ditto.

File: tests/forms/ViewModels/ItemsViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        public Command LoadItemsCommand { get; set; }
+
+        public ItemsViewModel()
+        {
+            Title = "Browse";
+            Items = new ObservableCollection<Item>();
+            LoadItemsCommand = new Command(async () => await ExecuteLoadItemsCommand());
+
+            MessagingCenter.Subscribe<NewItemPage, Item>(this, "AddItem", async (obj, item) =>
+            {
```
Comment (jonpryor) [inline tests/forms/ViewModels/ItemsViewModel.cs]: This `{` should be on the previous line.

File: tests/forms/ViewModels/ItemsViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    public class ItemsViewModel : BaseViewModel
+    {
+        public ObservableCollection<Item> Items { get; set; }
+        public Command LoadItemsCommand { get; set; }
+
+        public ItemsViewModel()
+        {
+            Title = "Browse";
+            Items = new ObservableCollection<Item>();
+            LoadItemsCommand = new Command(async () => await ExecuteLoadItemsCommand());
```
Comment (jonpryor) [inline tests/forms/ViewModels/ItemsViewModel.cs]: Space ` ` before `(`.

File: tests/forms/ViewModels/ItemsViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+            MessagingCenter.Subscribe<NewItemPage, Item>(this, "AddItem", async (obj, item) =>
+            {
+                var _item = item as Item;
+                Items.Add(_item);
+                await DataStore.AddItemAsync(_item);
+            });
+        }
+
+        async Task ExecuteLoadItemsCommand()
```
Comment (jonpryor) [inline tests/forms/ViewModels/ItemsViewModel.cs]: Space ` ` before `(`.

File: tests/forms/Views/MainPage.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        public MainPage()
+        {
+            Page itemsPage, aboutPage = null;
+
+            switch (Device.RuntimePlatform)
+            {
+                case Device.iOS:
+                    itemsPage = new NavigationPage(new ItemsPage())
+                    {
```
Comment (jonpryor) [inline tests/forms/Views/MainPage.cs]: This `{` should be on the previous line.

**Comment by jonpryor** [comment]
Overall, I don't like the name `tests/forms`. It doesn't have any semantic meaning, to answer "why does this directory exist?" (Compare to `tests/CodeGen-*`, which deals with code generation, or `tests/ResolveImports`, which deals with project resolution.) "forms" is a *what*, not a *why*.

I would prefer `tests/Xamarin.Forms-Performace-Integration`, or something otherwise semantically meaningful.

**Comment by radekdoulik** [comment]
rebased on current master, so hopefully it will build OK now.

regenerated the test with mono coding style set in VisualStudio/Mac, so hopefully it looks better now. also renamed the test directory and package name as suggested.

**Comment by jonpryor** [comment]
build

**Comment by jonpryor** [comment]
One downside to renaming `@(UnitTestApk)` to `@(TestApk)` is that this will require additional monodroid-side changes in order to merge. Is this desirable?

File: tests/Xamarin.Forms-Performance-Integration/Xamarin.Forms.Performance.Integration.sln [hunk 16]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (jonpryor) [inline tests/Xamarin.Forms-Performance-Integration/Xamarin.Forms.Performance.Integration.sln]: This file shouldn't exist. Instead, the .csproj should be added to the top level Xamarin.Android-Tests.sln file.

Comment (radekdoulik) [reply tests/Xamarin.Forms-Performance-Integration/Xamarin.Forms.Performance.Integration.sln]: Good catch, updated.

**Comment by radekdoulik** [comment]
The monodroid changes are simple, created https://github.com/xamarin/monodroid/pull/665 to do that. It can be completed with adding XA bump.

**Pull request merged**